native-lib: allow passing structs as arguments

nia-e opened the pull request:
Blocked on #4456 currently to make rebasing easier when that lands. This allows passing arbitrary structs as arguments to native code; receiving a struct as a return value is not yet implemented. Also includes a couple of tests, though we may want to add more (e.g. ensuring it's properly detecting UB when a non-`#[repr(C)]` struct is passed).

I figured the review queue looked a bit too empty :D

nia-e commented on tests/native-lib/fail/struct_not_extern_c.rs:
```diff
@@ -0,0 +1,19 @@
+
+fn main() {
+    let pass_me = PassMe { value: 42, other_value: 1337 };
+    unsafe { pass_struct(pass_me) }; //~ ERROR: Undefined Behavior: passing a non-#[repr(C)] struct over FFI
```
Is this correct? I think it's UB, but maybe if the user intentionally sets `#[allow(improper_ctypes)]` we might just want to report this as unsupported instead of UB?

nia-e commented:
Think this is ready for review. I split off a bunch of the new code this introduces into a different file to not pollute `native_lib/mod.rs` too much more than it already is

@rustbot ready

nia-e commented on src/shims/native_lib/mod.rs:
```diff
@@ -276,6 +279,115 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                let strukt = this.adt_to_carg(v.layout.ty, *adt_def, args)?;
+
+                // The raw bytes backing this arg.
+                let bytes = match v.as_mplace_or_imm() {
```
Can a struct ever be an imm? From my testing it seems like it's always an mplace here but I'm not sure

nia-e commented:
Should be fixed. Also bumped the libffi version since apparently there was a [soundness issue](https://github.com/libffi-rs/libffi-rs/commit/23726dca1f6e9164894978bc71e7229eb3b43c2a) that got resolved since the previous version we were on; that wasn't the cause of my test failures, but it's probably smart to include it anyways.

@rustbot ready

RalfJung commented on src/shims/native_lib/ffi.rs:
```diff
@@ -0,0 +1,113 @@
+    }
+}
+
+/// An owning form of `FfiArg`.
```
The prefixes "C" and "Ffi" don't do a great job reflecting "owned" and borrowed.

So what about `OwnedArg` and `BorrowedArg`? The "downcast" function (also an odd name, what does it downcast?) can then be called `borrow`.

`CPrimitive` could be called `ScalarArg` as that seems to match what we use it for.

Also, please generally define types bottom-up and with increasing complexity/subtlety, to simplify reading the file in order:
- ScalarArg
- OwnedArg
- BorrowedArg

RalfJung commented on src/shims/native_lib/ffi.rs:
```diff
@@ -0,0 +1,113 @@
+pub enum CArg {
+    /// Primitive type.
+    Primitive(CPrimitive),
+    /// Struct with its computed type layout and bytes.
+    Struct(FfiType, Box<[u8]>),
```
This doesn't have to be a struct, right? It can be any type that `FfiType` can represent, which includes basic integers.

nia-e replied:
I'll change the name, presumably the next PR idea which you brought up on Zulip would make this enum redundant anyway

RalfJung commented on tests/native-lib/pass/scalar_arguments.rs:
```diff
@@ -39,5 +39,54 @@ fn main() {
+    }
+}
+
+/// Test passing a basic struct as an argument.
+fn test_pass_struct() {
```
The file these tests are in is called "scalar_arguments". Do these arguments you are passing here ,ook like scalars to you? :)

Please make a new file for aggregate arguments.

RalfJung commented on src/shims/native_lib/mod.rs:
```diff
@@ -276,6 +278,147 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                        let ptr_raw = this.get_alloc_bytes_unchecked_raw(id)?;
+                        // SAFETY: We know for sure that at ptr_raw the next layout.size bytes
+                        // are part of this allocation and initialised. They might be marked as
+                        // uninit in Miri, but all bytes returned by the isolated allocator are
```
FWIW before https://github.com/rust-lang/miri/pull/4475 this argument was wrong on macOS where we didn't use the isolated allocator.

And in general there's no need to be so specific.
```suggestion
                        // uninit in Miri, but all bytes returned by `MiriAllocBytes` are
```

RalfJung commented on src/shims/native_lib/mod.rs:
```diff
@@ -276,6 +278,147 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+        args: &'tcx ty::List<ty::GenericArg<'tcx>>,
+    ) -> InterpResult<'tcx, FfiType> {
+        // TODO: is this correct? Maybe `repr(transparent)` when the inner field
+        // is itself `repr(c)` is ok?
```
Yes, that would be okay. But let's leave it to a future PR, since a `repr(transparent)` wrapper around e.g. `i32` should also be supported and that'll be a bit more interesting -- it's going to be an `Immediate`.

RalfJung commented on src/shims/native_lib/mod.rs:
```diff
@@ -276,6 +278,147 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+        }
+        // TODO: unions, etc.
+        if !adt_def.is_struct() {
+            throw_unsup_format!("unsupported argument type for native call: {orig_ty}");
```
```suggestion
            throw_unsup_format!("unsupported argument type for native call: {orig_ty} is an enum or union");
```

RalfJung reviewed (commented):
Thanks! libffi is much cooler than I expected. :)

I have some ideas for how we could unify the scalar and struct handling... but I think that's better done in a follow-up PR, so I'll post some things on Zulip.

@rustbot author

RalfJung commented on tests/native-lib/pass/scalar_arguments.rs:
```diff
@@ -39,5 +39,54 @@ fn main() {
+        part_2: Part2 { bits: 0xabcdef01 },
+        part_3: 0xabcdef01,
+    };
+    assert_eq!(unsafe { pass_struct_complex(complex) }, 0);
```
Can you add a fail test that looks roughly like this, to ensure we do check initialization properly?
```rust
let arg = MaybeUninit::<ComplexStruct>::uninit();
pass_struct_complex(*arg.as_ptr());
```
This needs `derive(Copy, Clone)` on the struct. (Don't use `.read()`, that might defeat the test.)

RalfJung commented on tests/native-lib/scalar_arguments.c:
```diff
@@ -30,6 +30,43 @@ EXPORT int64_t add_short_to_long(int16_t x, int64_t y) {
+} ComplexStruct;
+
+EXPORT int32_t pass_struct_complex(const ComplexStruct complex) {
+  if ((((uint32_t)complex.part_1.high) << 16 | (uint32_t)complex.part_1.low) == complex.part_2.bits
+      && complex.part_2.bits == complex.part_3)
```
Instead of doing some hard-to-follow shifts and whatever here, just make the function take 3 more arguments (all scalars) that contain the expected value for the 3 fields.

RalfJung commented on src/shims/native_lib/mod.rs:
```diff
@@ -276,6 +278,147 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                        // stripping provenance from them, since it'll later be
+                        // exposed recursively.
+                        let ptr = mplace.ptr();
+                        // Make sure the provenance of this allocation is exposed;
+                        // there must be one for this mplace to be valid at all.
+                        // The interpreter-level allocation itself is exposed in
+                        // visit_reachable_allocs.
+                        this.expose_provenance(ptr.provenance.unwrap())?;
```
Ah, I didn't understand the last sentence.

So yeah this will massively over-expose -- not just exposing the provenance of the struct itself is unnecessary, also consider that the struct might just span 10 bytes in a 1KB allocation and exposing all the other provenance in that allocation is equally unnecessary.

nia-e commented:
Apologies for the delay, IRL stuff came up, but this should be good now. I refactored the internals a little to do away with the need for a distinct borrowed/owned arg type, which should make the internal logic simpler, and added tests for a few more edge cases (notably the case of multiple structs in the same allocation; seems like I didn't need to change anything in rustc after all). Also rebased on main but all of the changes are in the last commit.

@rustbot ready

nia-e commented:
Hm, I rebased on master to make use of `get_range()` here as well and it seems like `repr(C)` structs *can* be immediates now (or at least, it causes tests to fail with that reason). Should I do the refactor you mentioned on Zulip in this branch? It's already mostly written locally

RalfJung commented on src/shims/native_lib/mod.rs:
```diff
@@ -267,6 +269,157 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                        let mplace_ptr = mplace.ptr();
+                        let sz = mplace.layout.size.bytes_usize();
+                        let id = this
+                            .alloc_id_from_addr(
+                                mplace_ptr.addr().bytes(),
```
Here you are doing the equivalent of casting a pointer to an integer and back. Please don't. You can get the alloc ID from the pointer directly.

RalfJung replied:
Specifically, please use `ptr_get_alloc_id`. That also gives you the offset inside the allocation.

However, you'll need to first separately handle the zero-sized case. `mplace.ptr()` might not point to an `Allocation` if the type has size 0.

nia-e replied:
Are ZSTs allowed over FFI? I was under the impression that there's no way to represent them in C so I figured just erroring here makes sense

nia-e commented:
Should be good. I couldn't use your suggestion of `write_target_uint` since that only takes a `u128` while a scalar pair could be up to 256 bits, so I had to adjust its logic to account for 2 scalars. Frankly that whole section could belong in a separate function to turn an immediate into bytes, but if one exists I couldn't find it.

I had also added some tests for that code (passing structs that were `a: i32, b: i16` and `a: i16, b: i32` so that there might be alignment padding between the fields) but removed them from here since `test_pass_struct` already handles the tricky case (second field with greater align than the first).

RalfJung commented:
> Should be good. I couldn't use your suggestion of write_target_uint since that only takes a u128 while a scalar pair could be up to 256 bits

You should be calling `write_target_uint` twice then, once for each scalar.

RalfJung commented on src/shims/native_lib/mod.rs:
```diff
@@ -267,6 +270,189 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                let mplace_ptr = mplace.ptr();
+                let sz = mplace.layout.size.bytes_usize();
+                if sz == 0 {
+                    throw_unsup_format!("Attempting to pass a ZST over FFI");
```
```suggestion
                    throw_unsup_format!("attempting to pass a ZST over FFI");
```

RalfJung commented on src/shims/native_lib/mod.rs:
```diff
@@ -267,6 +270,189 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                }
+            }
+            either::Either::Right(imm) => {
+                let (first, maybe_second) = imm.to_scalar_and_meta();
```
`to_scalar_and_meta` is meant for wide pointers, you don't have a wide pointer here. You only need to handle `ScalarPair` since `Scalar` should be truly impossible. It seems worth a little helper that deals with one scalar so that you can call it twice.

nia-e replied:
Sure, a `repr(C)` struct must be a scalar pair but this also needs to handle primitives, no? I can just match on the scalar directly instead of calling `to_scalar_and_meta()`, though

RalfJung replied:
> but this also needs to handle primitives, no?

Ah, is this also the codepath for regular scalars now? :+1: then.

RalfJung commented on src/shims/native_lib/mod.rs:
```diff
@@ -267,6 +270,189 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                    }
+                    MemPlaceMeta::None => (0, 0),
+                };
+                let sz = imm.layout.size.bytes_usize();
+                // TODO: Is this actually ok? Seems like the only way to figure
+                // out how the scalars are laid out relative to each other.
+                let align_second = match imm.layout.backend_repr {
+                    rustc_abi::BackendRepr::Scalar(_) => 1,
+                    rustc_abi::BackendRepr::ScalarPair(_, sc2) => sc2.align(this).bytes_usize(),
+                    _ => unreachable!(),
+                };
+                // How many bytes to skip between scalars if necessary for alignment.
+                let skip = sz_first.next_multiple_of(align_second).strict_sub(sz_first);
+
```
Have a look at https://github.com/rust-lang/rust/blob/ece1397e3ff40e7f8a411981844a8214659da970/compiler/rustc_const_eval/src/interpret/place.rs#L716-L732 for how to determine the offset of the 2nd field.

RalfJung commented on src/shims/native_lib/mod.rs:
```diff
@@ -267,6 +270,189 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                // give us an easy way to do multiple writes in a row, we
+                // adapt its logic for two consecutive writes.
+                let mut bytes_wr = bytes.as_mut();
+                match this.data_layout().endian {
+                    rustc_abi::Endian::Little => {
```
You should *not* be doing your own endianess handling. `write_target_uint` can do that for you.

RalfJung commented:
@rustbot author

nia-e commented:
@rustbot ready

nia-e commented on src/shims/native_lib/mod.rs:
```diff
@@ -267,6 +269,158 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                let write_scalar = |this: &MiriInterpCx<'tcx>, sc: Scalar, bytes: &mut [u8]| {
+                    // If a scalar is a pointer, then expose its provenance.
+                    if let interpret::Scalar::Ptr(p, _) = sc {
+                        // This relies on the `expose_provenance` in the `visit_reachable_allocs` callback
+                        // below to expose the actual interpreter-level allocation.
```
Is it wrong? If either scalar is a pointer, the code under the comment exposes provenance only in the AM; then later when `visit_reachable_allocs` is called it'll do `alloc.get_bytes_unchecked_raw().expose_provenance()` on the allocation said pointer points to. Unless I misunderstood the meaning of that comment, that's what it referred to in the first place

RalfJung replied:
It's correct that the code below only exposes the ptr in the AM. But the claim that this somehow relies on the ptr also being exposed in the "host" makes no sense. That referred to the `with_exposed_provenance_mut`, which as per its documentation requires a prior call to `expose_provenance`, and the comment was explaining where that `expose_provenance` happened.

RalfJung commented:
I've done some refactoring and cleanup, found it easier to just edit the code than write it out.^^ Please take a look at the commit I added and use that to train your model for future Miri PRs. ;)
